### ISSUE TITLE
ctf_sec - Blocklisted address or address(0) recipient can lock protocol fund when order is partially filled

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-dinari-judging#55]
ctf_sec

high

# Blocklisted address or address(0) recipient can lock protocol fund when order is partially filled

## Summary

Blocklisted address or address(0) recipient can lock protocol fund when order is partially filled

## Vulnerability Detail

When creating the order, the asset token and payment token is whitelited and validated

```solidity
   function requestOrder(OrderRequest calldata orderRequest, bytes32 salt) public nonReentrant whenOrdersNotPaused {
        // Reject spam orders
        if (orderRequest.quantityIn == 0) revert ZeroValue();
        // Check for whitelisted tokens
        _checkRole(ASSETTOKEN_ROLE, orderRequest.assetToken);
        _checkRole(PAYMENTTOKEN_ROLE, orderRequest.paymentToken);
        bytes32 orderId = getOrderIdFromOrderRequest(orderRequest, salt);
        // Order must not already exist
        if (_orders[orderId].remainingOrder > 0) revert DuplicateOrder();

        // Get order from request and move tokens
        Order memory order = _requestOrderAccounting(orderRequest, orderId);

        // Send order to bridge
        emit OrderRequested(orderId, order.recipient, order, salt);

        // Initialize order state
        uint256 orderAmount = order.sell ? order.assetTokenQuantity : order.paymentTokenQuantity;
        _orders[orderId] = OrderState({requester: msg.sender, remainingOrder: orderAmount, received: 0});
        _numOpenOrders++;
    }

```

but the order.recipient is never validated

the payment token can be USDC and support blocklist, the user can set order.recipient to a blocked token address

the user can also set the recipient address to address(0)

then later when the operator want to fill in the order

when order is partially filled, the transaction proceeded, the share is burned and the payment token is transferred in the OrderProcessor contract

```solidity

    /// @inheritdoc OrderProcessor
    function _fillOrderAccounting(
        OrderRequest calldata orderRequest,
        bytes32 orderId,
        OrderState memory orderState,
        uint256 fillAmount,
        uint256 receivedAmount
    ) internal virtual override {
        // Accumulate fee obligations at each sill then take all at end
        uint256 collection = getPercentageFeeForOrder(receivedAmount);
        uint256 feesEarned = _feesEarned[orderId] + collection;
        // If order completely filled, clear fee data
        uint256 remainingOrder = orderState.remainingOrder - fillAmount;
        if (remainingOrder == 0) {
            // Clear fee state
            delete _feesEarned[orderId];
        } else {
            // Update fee state with earned fees
            if (collection > 0) {
                _feesEarned[orderId] = feesEarned;
            }
        }

        // Burn asset
        IMintBurn(orderRequest.assetToken).burn(fillAmount);
        // Transfer raw proceeds of sale here
        IERC20(orderRequest.paymentToken).safeTransferFrom(msg.sender, address(this), receivedAmount);
        // Distribute if order completely filled
        if (remainingOrder == 0) {
            _distributeProceeds(
                orderRequest.paymentToken, orderRequest.recipient, orderState.received + receivedAmount, feesEarned
            );
        }
    }
```

note the code section:

```solidity
   // Burn asset
        IMintBurn(orderRequest.assetToken).burn(fillAmount);
        // Transfer raw proceeds of sale here
        IERC20(orderRequest.paymentToken).safeTransferFrom(msg.sender, address(this), receivedAmount);
        // Distribute if order completely filled
        if (remainingOrder == 0) {
            _distributeProceeds(
                orderRequest.paymentToken, orderRequest.recipient, orderState.received + receivedAmount, feesEarned
            );
        }
```

only if the remaining order is 0, meaning when the order is fully filled, the payment token is distributed

However, even when the order is filled, and when calling _distributeProceeds, transaction still revert because the order.recipient is a blocklisted addres or address(0)

```solidity
   function _distributeProceeds(address paymentToken, address recipient, uint256 totalReceived, uint256 feesEarned)
        private
    {
        // Check if accumulated fees are larger than total received
        uint256 proceeds = 0;
        uint256 collection = 0;
        if (totalReceived > feesEarned) {
            // Take fees from total received before distributing
            proceeds = totalReceived - feesEarned;
            collection = feesEarned;
        } else {
            // If accumulated fees are larger than total received, then no proceeds go to recipient
            collection = totalReceived;
        }

        // Transfer proceeds to recipient
        if (proceeds > 0) {
            IERC20(paymentToken).safeTransfer(recipient, proceeds);
        }
        // Transfer fees to treasury
        if (collection > 0) {
            IERC20(paymentToken).safeTransfer(treasury, collection);
        }
    }
```

transaction revert when calling this [line of code](https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/SellOrderProcessor.sol#L171)

```solidity
 IERC20(paymentToken).safeTransfer(recipient, proceeds);
```

there is just no way for protocol to get the fund back because even when canceling the order,

 transaction still revert in this [line of code](https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/SellOrderProcessor.sol#L171)
 because _distributeProceeds when there are partially fill order
 
```solidity
    /// @inheritdoc OrderProcessor
    function _cancelOrderAccounting(OrderRequest calldata orderRequest, bytes32 orderId, OrderState memory orderState)
        internal
        virtual
        override
    {
        // If no fills, then full refund
        uint256 refund;
        if (orderState.remainingOrder == orderRequest.quantityIn) {
            // Full refund
            refund = orderRequest.quantityIn;
        } else {
            // Otherwise distribute proceeds, take accumulated fees, and refund remaining order
            _distributeProceeds(
                orderRequest.paymentToken, orderRequest.recipient, orderState.received, _feesEarned[orderId]
            );
            // Partial refund
            refund = orderState.remainingOrder;
        }

        // Clear fee data
        delete _feesEarned[orderId];

        // Return escrow
        IERC20(orderRequest.assetToken).safeTransfer(orderRequest.recipient, refund);
    }
```

note, if this line of code:

```solidity
  // Return escrow
        IERC20(orderRequest.assetToken).safeTransfer(orderRequest.recipient, refund);
```

if the orderRequest.recipient address is a blocklisted address by dShares transferRestrictor, the transaction revert as well, blocking order canceling.
 
## Impact

Blocklisted address or address(0) recipient can lock protocol fund when order is partially filled

## Code Snippet

https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/BuyOrderIssuer.sol#L214

https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/SellOrderProcessor.sol#L171

## Tool used

Manual Review

## Recommendation

We recommend the protocol validate the recipient address to not let user set a blocklisted address or address(0)

**Comment by Delvir0** [comment]
Escalate

1. Dup of https://github.com/sherlock-audit/2023-06-dinari-judging/issues/57, both issues even mention the exact same recommendation "We recommend the protocol validate the recipient address to not let user set a blocklisted address or address(0)". Meaning the orgin of issue is the same.
2. Sherlock criteria for validity is as follows: 
![image](https://github.com/sherlock-audit/2023-06-dinari-judging/assets/111504898/812465f6-3f45-40a0-b044-abcd024a2d8d)
Being blacklisted while the impact is only on the blacklisted user != valid
3. inputting address(0) as recipient is an user error, meaning not valid:
![image](https://github.com/sherlock-audit/2023-06-dinari-judging/assets/111504898/cf5c745e-d3b3-495a-a47c-5e8a927d4bcd)
4. If the user is de-blocked, the refund can still happen since the order still exists since it can't be cancelled.
5. At most the cause can be loss of gas to the operator which can be a on time thing since the off-chain mechanism is centralized and either canceling or ordering can be cancelled for a given orderId.

**Comment by sherlock-admin** [comment]
> Escalate
> 
> 1. Dup of https://github.com/sherlock-audit/2023-06-dinari-judging/issues/57, both issues even mention the exact same recommendation "We recommend the protocol validate the recipient address to not let user set a blocklisted address or address(0)". Meaning the orgin of issue is the same.
> 2. Sherlock criteria for validity is as follows: 
> ![image](https://github.com/sherlock-audit/2023-06-dinari-judging/assets/111504898/812465f6-3f45-40a0-b044-abcd024a2d8d)
> Being blacklisted while the impact is only on the blacklisted user != valid
> 3. inputting address(0) as recipient is an user error, meaning not valid:
> ![image](https://github.com/sherlock-audit/2023-06-dinari-judging/assets/111504898/cf5c745e-d3b3-495a-a47c-5e8a927d4bcd)
> 4. If the user is de-blocked, the refund can still happen since the order still exists since it can't be cancelled.
> 5. At most the cause can be loss of gas to the operator which can be a on time thing since the off-chain mechanism is centralized and either canceling or ordering can be cancelled for a given orderId.

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by ctf-sec** [comment]
> Escalate
> 
> 1. Dup of [ctf_sec - Blocklisted recipient or address(0) force operator to waste gas when filling the sell order or filling the cancel order #57](https://github.com/sherlock-audit/2023-06-dinari-judging/issues/57), both issues even mention the exact same recommendation "We recommend the protocol validate the recipient address to not let user set a blocklisted address or address(0)". Meaning the orgin of issue is the same.
> 2. Sherlock criteria for validity is as follows:
>    ![image](https://user-images.githubusercontent.com/111504898/252648150-812465f6-3f45-40a0-b044-abcd024a2d8d.png)
>    Being blacklisted while the impact is only on the blacklisted user != valid
> 3. inputting address(0) as recipient is an user error, meaning not valid:
>    ![image](https://user-images.githubusercontent.com/111504898/252649645-cf5c745e-d3b3-495a-a47c-5e8a927d4bcd.png)
> 4. If the user is de-blocked, the refund can still happen since the order still exists since it can't be cancelled.
> 5. At most the cause can be loss of gas to the operator which can be a on time thing since the off-chain mechanism is centralized and either canceling or ordering can be cancelled for a given orderId.

The user can intentionally use address(0) as recipient and any protocol fund that used to partially filled the order is locked forever. The impact is more much severe than waste gas, not a duplicate of #57

**Comment by Delvir0** [comment]
Don't see how protocol funds are locked. Users sends the amount in USDC for the orders when requesting an order. 
When it's filled the fillAmount required for the stock purchase is sent to the operator. Meaning the scope of the **potential** loss is for the user. 

Also, since filling the order reverts, the funds of the users stay in the contract until an unblock is triggered and cancelOrder is called.

Furthermore, this issue and #57  have the same "problem" but simply note a different kind of impact which all regard the same scenario.

**Comment by JeffCX** [comment]
Yes, exactly the issue and #57 has the same problem but the impact is more severe in this finding combing with partial fill.

> Don't see how protocol funds are locked. Users sends the amount in USDC for the orders when requesting an order.
When it's filled the fillAmount required for the stock purchase is sent to the operator. Meaning the scope of the potential loss is for the user.

this is buy order, correct

when selling the order, user needs to offer dShares in exchange for payment token (USDC), any partially filled USDC (protocol fund) is locked, thought the original report is clear on this.

**Comment by Delvir0** [comment]
"any partially filled USDC (protocol fund)" this is not true. That amount of USDC is entitled to the user since he sold his shares.

This was my last comment, leaving it to the judge to not further spam this.

**Comment by ctf-sec** [comment]
User cannot receive the USDC payment token, we are assume the malicious user does not want to receive payment token if he specify an address(0) as recipient.

In normal case, in the worst case the protocol can forcefully call cancel order to retrieve the partially filled payment token.

In this case, protocol can't either because order are not canceled-> loss of partially filled fund.

**Comment by bizzyvinci** [comment]
No protocol funds are lost. Protocol funds are instantly transferred when filling buy and sell orders.

Transferred TO protocol for buy orders (cause user paid while creating order)
https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/BuyOrderIssuer.sol#L156

Transferred FROM protocol for sell order (so that user can have the token when everything is filled)
https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/SellOrderProcessor.sol#L116


I believe ctf-sec is talking about treasury funds [here](https://github.com/sherlock-audit/2023-06-dinari/blob/4851cb7ebc86a7bc26b8d0d399a7dd7f9520f393/sbt-contracts/src/issuer/SellOrderProcessor.sol#L175). But for address banned by USDC or 0x0 which should not have been filled in the first place as suggested by ctf-sec, then the treasury doesn't deserve fee for such a transaction.

Blacklist and input validation only protects the user, not the protocol.

**Comment by ctf-sec** [comment]
**In buy order, user use USDC to buy dShares**

so USDC is considered user's fund and dShares is considered protocol's fund

Not really a issue in Buy Order

In Sell Order, user use dShares and sell dShares for USDC

In this case, the dShares is user's fund and the USDC payment token is protocol's fund.

**Not sure why escalation think in sell order the payment token is the user's fund....**

The root cause + impact present in the original report states: "partially filled order trap protocol's fund" is still valid given the code logic

there seems like a disagreement in the assumption that if the operator offchain should validate the recipient address before filling the order

> But for address banned by USDC or 0x0 which should not have been filled in the first place 

how the operator do this seems not in the scope of the auditing.

The operator is considered as trusted party and a very reasonable assumption is the operator filling the order with decent amount at a reasonable price level

If operator do not filling the order, this defeats the purpose of using the protocol......

There are two check that can be done by operator:

- simulate the partially filled transaction to make sure the transaction not revert before sending transaction to partially fill the order
- this check would pass because the partially filled order would not revert, it only involves sending protocol's payment token to the contract

- simulate and implement the logic to validate the recipient address is not blacklisted or address(0)
- this does not really help either

consider the case when a address is not blacklisted, after partially filled, the address is blacklisted, the protocol's payment that can be used to fill other order is still trapped

**again, how does the operator do and whether operator validate the blacklisted recipient is not in scope of the auditing, assuming operator would faithfully partially fill order is a very reasonable to make when auditing.**

The fix is not pushing the order out and let user claim the fund and if after a expiry period user do not claim the fund, the operator help user claim the fund to fundamentally resolve this issue

![image](https://github.com/sherlock-audit/2023-06-dinari-judging/assets/114844362/a053a274-2931-434b-b78b-94a36477a12f)

there is no mention about the receiver input validation as well

**Comment by bizzyvinci** [comment]
> In Sell Order, user use dShares and sell dShares for USDC

Yes, absolutely true and that's the main point.

> In this case, the dShares is user's fund and the USDC payment token is protocol's fund.

False
* There's no USDC in the contract when order has not been filled
* For each filled order, dShares would be burn, and **protocol would transfer USDC instantly for the user** (which would be left in the contract until full filling or cancelling)

dShares and USDC in the contract belongs to the user. dShares are Shares that have not been filled, while the USDC is the earning from the filled orders.

For example, a user wants to sell 5 dShares at a price of 1000USDC
* Before filling, he has 5dShares
* After filling 1 dShares, he would have 4=5-1 dShares, and the protocol would transfer 1000USDC to the contract.
* After filling another 2 dShares, he would have 2=4-2 dShares, and the protocol would transfer another 2000USDC
Now, the contract would have 2 dShares belonging to the user for order that wasn't filled, and 3000USDC for the 3 dShares that were filled.

**Comment by Oot2k** [comment]
To summarize:

User loses the most, all his burned shares and usdc
Protocol loses Treasury fees
The operator loses gas and other costs, which are out of scope.

I think this and the break of core functionality are enough to mark this as a medium severity issue.

Compared to issue #57 the impact of this issue is much more severe, and in the past, there have been cases where issues similar to this were accepted as high or medium.

**Comment by Delvir0** [comment]
@Oot2k just a quick note according to your summarization. 
Taken the in scope scenario's, which are indeed (a) User loses the most, all his burned shares and usdc (b) Protocol loses Treasury fees. 
Side note: the core functionalities for all other users remain in operation and non is halted (except for this specific order for the specific user). 

The attack vector which the Watson mentions is a grieving attack which makes the protocol lose gas and other costs (out of scope) and perhaps Treasury fees (which would be around 1 wei anyway since the attack would not include a lot since the attacker will lose it)

Furthermore watson mentions:

"but the order.recipient is never validated
the payment token can be USDC and support blocklist, the user can set order.recipient to a blocked token address
the user can also set the recipient address to address(0)"

Meaning two conclusions can be taken when the user performs one of the two actions above:
1. In case of (a), this does not exist since it's an attack. If, let's say the user submitted a legit order with address(0) on accident, that would be an user error which is invalid
2. In case of (b) the impact would be minimal

I do agree on this needing to be fixed. Hence a valid low.

**Comment by jaketimothy** [comment]
Fixed in
- https://github.com/dinaricrypto/sbt-contracts/pull/116

**Comment by hrishibhat** [comment]
@ctf-sec 
Based on the above comments and this one
https://github.com/sherlock-audit/2023-06-dinari-judging/issues/55#issuecomment-1634173542
It seems like the attacker has to **lose the funds owed to them** in order to cause: 
- gas loss for operator 
- and treasury fees for protocol. 
Is that correct?

**Comment by ctf-sec** [comment]
> @ctf-sec Based on the above comments and this one [#55 (comment)](https://github.com/sherlock-audit/2023-06-dinari-judging/issues/55#issuecomment-1634173542) It seems like the attacker has to **lose the funds owed to them** in order to cause:
> 
> * gas loss for operator
> * and treasury fees for protocol.
>   Is that correct?

the fund that is used to partially fill the order is a lot more than treausry fee

**Comment by ctf-sec** [comment]
> @ctf-sec Based on the above comments and this one [#55 (comment)](https://github.com/sherlock-audit/2023-06-dinari-judging/issues/55#issuecomment-1634173542) It seems like the attacker has to **lose the funds owed to them** in order to cause:
> 
> * gas loss for operator
> * and treasury fees for protocol.
>   Is that correct?

This is correct, 

I have to leave factual based comments, in the current version of the code, 

there is no way to get the partially filled USDC back from protocol

```solidity
  function _distributeProceeds(address paymentToken, address recipient, uint256 totalReceived, uint256 feesEarned)
        private
    {
        // Check if accumulated fees are larger than total received
        uint256 proceeds = 0;
        uint256 collection = 0;
        if (totalReceived > feesEarned) {
            // Take fees from total received before distributing
            proceeds = totalReceived - feesEarned;
            collection = feesEarned;
        } else {
            // If accumulated fees are larger than total received, then no proceeds go to recipient
            collection = totalReceived;
        }

        // Transfer proceeds to recipient
        if (proceeds > 0) {
            IERC20(paymentToken).safeTransfer(recipient, proceeds);
        }
        // Transfer fees to treasury
        if (collection > 0) {
            IERC20(paymentToken).safeTransfer(treasury, collection);
        }
    }
```

the code above means the partially filled USDC goes to recipient (revert), the fee goes to treasury

**Comment by ctf-sec** [comment]
> Fixed in
> 
> * [Feat/checker zero and block list addresses dinaricrypto/sbt-contracts#116](https://github.com/dinaricrypto/sbt-contracts/pull/116)

https://github.com/dinaricrypto/sbt-contracts/blob/fc3fb85236ac5bfbe1bfa1e8f0a2d07ad270658f/src/issuer/OrderProcessor.sol#L262

Emm Recommend validating:

1. the recipient address is not address(0) when requesting order
2. validating the order requestor and order recipient is not blocklisted by both protocol and the payment token admin

the input validation in the RequestOrder section does not resolve the issue when user address is blocklisted halfway and protocol cannot get the fee

I would recommend adding some code like:

1. if transferring fund to the user failed, the user can claim the fund himself within a certain time window (like a WEEK)
2. after the time windows, protocol can help user sweep the fund to avoid locking the payment token

**Comment by hrishibhat** [comment]
Result:
Low
Has duplicates
After considering all the information above and further discussing with the protocol considering this issue a valid low. 
Sponsor comment:
> If a sell order was partially filled, then the recipient is blacklisted on USDC (note an initial address check would not be able to catch this) the order would be locked. Fill/fulfill and cancel will both revert. The user loses access to the USDC proceeds from earlier fill (which maybe they could have moved to different wallet before lock). Protocol loses fees for earlier fills.

 - In case of the deliberate lock the user would have to forego the USDC proceeds. 
So their loss would be much larger than loss of the protocol fee.

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [Delvir0](https://github.com/sherlock-audit/2023-06-dinari-judging/issues/55/#issuecomment-1630652983): accepted

**Comment by jaketimothy** [comment]
Additional blacklist checks in
- https://github.com/dinaricrypto/sbt-contracts/pull/164

Sending fees and proceeds per fill to prevent locked tokens in
- https://github.com/dinaricrypto/sbt-contracts/pull/161

**Comment by ctf-sec** [comment]
This fix changed all files, can be considered as a new feature instead of in the fix review scope

**Comment by MLON33** [comment]
Confirming that this fix was not in scope for the Sherlock fix review.
